src/table/boot.rs

```diff
@@ -203,6 +205,43 @@ impl BootServices {
         (self.free_pool)(addr).into()
     }
 
+    /// Stops execution until an event is signaled
+    ///
+    /// This function must be called at priority level Tpl::Application. If an
+    /// attempt is made to call it at any other priority level, an `Unsupported`
+    /// error is returned.
+    ///
+    /// The input Event slice is repeatedly iterated from first to last until an
+    /// event is signaled or an error is detected. The following checks are
+    /// performed on each event:
+    ///
+    /// * If an event is of type NotifySignal, then an `InvalidParameter` error
+    ///   is returned together with the index of the event that caused the failure.
+    /// * If an event is in the signaled state, the signaled state is cleared
+    ///   and the index of the event that was signaled is returned.
+    /// * If an event is not in the signaled state but does have a notification
+    ///   function, the notification function is queued at the event's
+    ///   notification task priority level. If the execution of the event's
+    ///   notification function causes the event to be signaled, then the
+    ///   signaled state is cleared and the index of the event that was signaled
+    ///   is returned.
+    ///
+    /// To wait for a specified time, a timer event must be included in the
+    /// Event slice.
+    ///
+    /// To check if an event is signaled without waiting, an already signaled
+    /// event can be used as the last event in the slice being checked, or the
+    /// check_event() interface may be used.
+    pub fn wait_for_event(&self, events: &mut [Event]) -> result::Result<usize, (Status, usize)> {
+        let (number_of_events, events) = (events.len(), events.as_mut_ptr());
+        let mut index = unsafe { mem::uninitialized() };
+        match (self.wait_for_event)(number_of_events, events, &mut index) {
+            Status::Success => Ok(index),
+            s @ Status::InvalidParameter => Err((s, index)),
+            error => Err((error, 0)),
```

Comment:
An alternative to this gratuitous zero would be an enumerated error type which has a "NotifySignalFailes" alternative with an index and an "OtherError" alternative without. With a bit of extra run-time checking, this could also be used to better differentiate between the two different meanings of InvalidParameter in this function.

Reply:
The idea is that `events` being empty and a `NotifySignal` being present in the list shouldn't happen in a well-coded production app, so the `InvalidParameter` error in this case is more of a helper during development.

Same for `Tpl::Application`, this could only happen during development, if the developer is trying to call this function from a callback function which is already servicing an event. So it might even make sense to add an `assert!` here, since a real app wouldn't really have a meaningful way to handle an error.

As for any other error, does the UEFI spec say that the implementation is allowed to return a different error code from the ones listed in the spec? I mean, in theory this function can only return Success, InvalidParameter, or Unsupported. Any other error should be marked as `unreachable!`.

Reply:
The UEFI spec provides no guarantee about completeness of the status code lists. In fact, firmware-specific status codes are slightly encouraged by this bit on protocols (2.7A p. 46):

> Any status codes defined by the protocol member function definition are required to be implemented, **Additional error codes may be returned**, but they will not be tested by standard compliance tests, and any software that uses the procedure cannot depend on any of the extended error codes that an implementation may provide.

Come to think of it, this means that a Rust enum might not be the right tool to model UEFI status code, as putting an unexpected value in a Rust enum [is undefined behaviour](https://doc.rust-lang.org/reference/behavior-considered-undefined.html).

Reply:
Ouch, this is a pretty big problem. Not sure what the most ergonomic way of handling this would be. 
Would forcing people to use [non-exhaustive pattern matching](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md) help?

Reply:
It could be part of the answer, but it does not by itself resolve the enum UB issue.

Maybe we should be making Status a newtype of usize and define constants that match the current enum variants. It's ugly, but it works without being UB, and it also allows users to match firmware-specific status codes which we don't support.

Reply:
I've opened #48 to track this further.